File: leetcode2/StickerstoSpellWord/a.cpp
/* Copyright (C) 2020 by iamslash */

#include <cstdio>
#include <unordered_map>
#include <vector>

using namespace std;

//     stickers: with example science
//       target: thehat
// stickerFreqs: i h t w    a e l m p x    c e i n s
//               1 1 1 1    1 2 1 1 1 1    2 2 1 1 1
//   targetFreq: a e h t
//               1 1 2 2

// 68ms 93.97% 12.3MB 48.23%
// dynamic programming
// O(2^T*S*T) O(2^T)
class Solution {
  int dfs(unordered_map<string, int>& C,
          vector<vector<int>>& stickerFreqs, string& target) {
    // base
    if (target.empty())
      return 0;
    // memo
    if (C.count(target))
      return C[target];
    // recursion
    int ans = INT_MAX, n = stickerFreqs.size();
    vector<int> targetFreq(26, 0);
    for (char  c : target)
      targetFreq[c-'a']++;
    for (int i = 0; i < n; ++i) {
      if (stickerFreqs[i][target[0]-'a'] == 0)
        continue;
      string s;
      for (int j = 0; j < 26; ++j) {
        if (targetFreq[j] - stickerFreqs[i][j] > 0) {
          s += string(targetFreq[j] - stickerFreqs[i][j], 'a' + j);
        }
      }
      int tmp = dfs(C, stickerFreqs, s);
      if (tmp != -1) {
        ans = min(ans, 1 + tmp);
      }        
    }
    C[target] = ans == INT_MAX? -1 : ans;
    return C[target];
  }
 public:
  int minStickers(vector<string>& stickers, string target) {
    int n = stickers.size();
    vector<vector<int>> stickerFreqs(n, vector<int>(26, 0));
    unordered_map<string, int> C;
    for (int i = 0; i < n; ++i) {
      for (char c : stickers[i]) {
        stickerFreqs[i][c-'a']++;
      }
    }
    return dfs(C, stickerFreqs, target);
  }
};

File: leetcode2/StickerstoSpellWord/a.py
# -*- coding: utf-8 -*-
# Copyright (C) 2024 by iamslash

from collections import Counter, defaultdict
from typing import List

# 263ms 73.32% 16.6MB 98.77%
# recursive dynamic programming
# O(N * 2^M) O(2^M + N)
# N: stickers count
# M: target length 
class Solution:
    
    def dfs(self, cache, sticker_freqs, target):
        # base
        if not target:
            return 0
        
        # memo
        if target in cache:
            return cache[target]
        
        # recursion
        target_freq = Counter(target)

        min_cnt = float('inf')
        for sticker_freq in sticker_freqs:
            if sticker_freq[target[0]] == 0:
                continue

            new_target_freq = target_freq - sticker_freq            
            new_target = ''.join(char * count for char, count in new_target_freq.items())

            if new_target != target:
                cnt = self.dfs(cache, sticker_freqs, new_target)
                if cnt != -1:
                    min_cnt = min(min_cnt, cnt + 1)

        cache[target] = -1 if min_cnt == float('inf') else min_cnt
        return cache[target]
    
    def minStickers(self, stickers: List[str], target: str) -> int:
        sticker_freqs = [Counter(sticker) for sticker in stickers]
        cache = {}
        return self.dfs(cache, sticker_freqs, target)
